### ISSUE TITLE
[MACHINE LEARNING] Using C++ to build a model for your Python Machine Learning Project

### ISSUE THREAD

**Issue by Agusioma** [section-engineering-education/engineering-education#6006]
### Topic Suggestion
[MACHINE LEARNING] Using C++ to build a optimized model for your Python Machine Learning Project

### Writing sample(s): 
My latest sample is found [here](https://www.section.io/engineering-education/android-jni-cpluplus-dijkstra/)

#### Proposed title of article
Using C++ to build a model for your Python Machine Learning Project

#### Proposed article introduction
It's with no doubt that Python is more versatile in building Machine Learning models. This is most definitely due to the large community, many libraries, short and easy-to-understand code, etc. However, it has its shortfalls when it comes to execution speed. This is where a high-speed language like C++ comes in. We can build a fast executing ML model using C++, but it's no match to Python in the number of Machine Learning libraries. How about utilizing Python libraries such as Numpy and Pandas for data preprocessing and then building a model running on C++? That's what we're going to look at.

Luckily, Python has the **ctypes** module that allows us to call C++ code and use it in Python code. In this article, we are going to do exactly that. We will use C++ to build a Logistic Regression model and then regularize it using Gradient Descent.

#### Key takeaways 
The user will learn about:
- Python's **ctypes** module.
- Building a machine learning model from scratch using C++.
- Consuming the C++ model using Python.

#### Article quality
It will be easy and simple to follow. Everything will be done from scratch and broken down bit by bit.

#### References
[Towards Data Science](https://towardsdatascience.com/calling-c-code-from-python-with-ctypes-module-58404b9b3929)

**Comment by the-confused-genius** [comment]
Hmm seems a good idea.
I have built a program on LinearRegression using c, which is more accurate and many times faster than python scripts.

**Comment by Agusioma** [comment]
Thank you so much. I am finalizing on my research, and I'll upload it
probably Saturday. Could you share with me the link you did on Linear
regression?

**Comment by the-confused-genius** [comment]
I have attached the link to my file on linear regression using c
https://github.com/the-confused-genius/Best-fit-curve/tree/main/Linear%20Regression

Using this we could find the best fit line for any linear model. With the minimum input of two points. The output of the program is m which is the slop and c which is the intercept of the line.

**Comment by Agusioma** [comment]
> I have attached the link to my file on linear regression using c https://github.com/the-confused-genius/Best-fit-curve/tree/main/Linear%20Regression
> 
> Using this we could find the best fit line for any linear model. With the minimum input of two points. The output of the program is m which is the slop and c which is the intercept of the line.

Alright, I'll check it out.

**Comment by Agusioma** [comment]
> ### Topic Suggestion
> [MACHINE LEARNING] Using C++ to build a model for your Python Machine Learning Project
> 
> ### Writing sample(s):
> My latest sample is found [here](https://www.section.io/engineering-education/android-jni-cpluplus-dijkstra/)
> 
> ### Proposal Submission
> #### Proposed title of article
> Using C++ to build a model for your Python Machine Learning Project
> 
> #### Proposed article introduction
> It's with no doubt that Python is more versatile in building Machine Learning models. This is most definitely due to the large community, many libraries, short and easy-to-understand code, etc. However, it has its shortfalls when it comes to execution speed. This is where a high-speed language like C++ comes in. We can build a fast executing ML model using C++, but it's no match to Python in the number of Machine Learning libraries. How about utilizing Python libraries such as Numpy and Pandas for data preprocessing and then building a model running on C++? That's what we're going to look at.
> 
> Luckily, Python has the **ctypes** module that allows us to call C++ code and use it in Python code. In this article, we are going to do exactly that. We will use C++ to build a Logistic Regression model and then regularize it using Gradient Descent.
> 
> #### Key takeaways
> The user will learn about:
> 
> * Python's **ctypes** module.
> * Building a machine learning model from scratch using C++.
> * Consuming the C++ model using Python.
> 
> #### Article quality
> It will be easy and simple to follow. Everything will be done from scratch and broken down bit by bit.
> 
> #### References
> [Towards Data Science](https://towardsdatascience.com/calling-c-code-from-python-with-ctypes-module-58404b9b3929)

**Comment by Agusioma** [comment]
@the-confused-genius, it was published yesterday. [Here](https://www.section.io/engineering-education/cpp-model-in-a-python-machine-learning-project/) it is. Though I tried passing a CSV directly from Python I did not succeed in that(I tried for close to 3 consecutive days). The pointer manipulation and memory allocations need more research. I will do that on a dedicated article once I finish my research. For testing, I just used a sample 2D array I extracted from a heart data CSV file. It was worth it.
